plugins/embed-optimizer/class-embed-optimizer-tag-visitor.php

```diff
@@ -283,4 +195,134 @@ private function reduce_layout_shifts( OD_Tag_Visitor_Context $context ): void {
+			// https://lf16-cdn-tos.tiktokcdn-us.com, and https://lf16-tiktok-common.tiktokcdn-us.com among others
+			// which either appear to be geo-targeted ('-us') _or_ load-balanced ('lf16'). So these are not added
+			// to the preconnected hosts.
+		} elseif ( $has_class( 'wp-block-embed-amazon' ) ) {
+			$preconnect_hrefs[] = 'https://read.amazon.com';
+			$preconnect_hrefs[] = 'https://m.media-amazon.com';
+		} elseif ( $has_class( 'wp-block-embed-soundcloud' ) ) {
+			$preconnect_hrefs[] = 'https://w.soundcloud.com';
+			$preconnect_hrefs[] = 'https://widget.sndcdn.com';
+			// Note: There is also https://i1.sndcdn.com which is for the album art, but the '1' indicates it may be geotargeted/load-balanced.
+		} elseif ( $has_class( 'wp-block-embed-pinterest' ) ) {
+			$preconnect_hrefs[] = 'https://assets.pinterest.com';
+			$preconnect_hrefs[] = 'https://widgets.pinterest.com';
+			$preconnect_hrefs[] = 'https://i.pinimg.com';
+		}
+
+		return $preconnect_hrefs;
+	}
+
+	/**
+	 * Adds preconnect links for embed resources.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context, with the cursor currently at an embed block.
+	 */
+	private function add_preconnect_links( OD_Tag_Visitor_Context $context ): void {
+		$processor           = $context->processor;
+		$embed_wrapper_xpath = self::get_embed_wrapper_xpath( $processor->get_xpath() );
+
+		$preconnect_hrefs = $this->get_preconnect_urls( $processor );
+		foreach ( $preconnect_hrefs as $preconnect_href ) {
+			foreach ( $context->url_metric_group_collection as $group ) {
+				if ( ! ( $group->get_element_max_intersection_ratio( $embed_wrapper_xpath ) > 0.0 ) ) {
+					continue;
+				}
+
+				$context->link_collection->add_link(
+					array(
+						'rel'  => 'preconnect',
+						'href' => $preconnect_href,
+					),
+					$group->get_minimum_viewport_width(),
+					$group->get_maximum_viewport_width()
+				);
+			}
+		}
+	}
+
+	/**
+	 * Optimizes an embed based on whether it is displayed in any initial viewport.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context, with the cursor currently at an embed block.
+	 */
+	private function lazy_load_embeds( OD_Tag_Visitor_Context $context ): void {
+		$processor = $context->processor;
+
+		// Lazy-loading can only be done once there are URL Metrics collected for both mobile and desktop.
+		if (
+			$context->url_metric_group_collection->get_first_group()->count() === 0
+			||
+			$context->url_metric_group_collection->get_last_group()->count() === 0
+		) {
+			return;
+		}
+
+		$embed_wrapper_xpath = self::get_embed_wrapper_xpath( $processor->get_xpath() );
+
+		$max_intersection_ratio = $context->url_metric_group_collection->get_element_max_intersection_ratio( $embed_wrapper_xpath );
+		if ( ! ( $max_intersection_ratio > 0.0 ) && embed_optimizer_update_markup( $processor, false ) && ! $this->added_lazy_script ) {
```

Comment:
@westonruter Do we need to use `PHP_FLOAT_EPSILON` here? Wouldn't `! ( $max_intersection_ratio > 0.0 )` suffice, as that's what is used when adding preconnect links?

Reply:
I think `! ( $max_intersection_ratio > 0.0 )` is equivalent to `0.0 === $max_intersection_ratio`, and both seem to work fine, but it seems in some cases even when a finite decimal is provided, it can result in a very unexpected value, for example #1765. So in this way, it seems safer to do:

```suggestion
		if ( $max_intersection_ratio < PHP_FLOAT_EPSILON && embed_optimizer_update_markup( $processor, false ) && ! $this->added_lazy_script ) {
```

Reply:
Makes sense, thanks for clarifying. Done in 0b4150c.